### DISCUSSION TITLE
PlotCurveItem with connect='finite' no longer works when the x/y values contain nan or inf.

### DISCUSSION THREAD

**Discussion by proptrader33** [pyqtgraph/pyqtgraph#2290] [General]
Hello,

I noticed my PlotCurveItem no longer works correctly.

I insert numpy.nan into the xy arrays for my PlotCurveItems where I need to disconnect the line at specific points. And I do believe this is what the "connect='finite'" property is meant to be used for.

With the new release, however, this feature no longer works and when the xs or ys array contains a nan or inf value, the whole item does not get plotted on the chart, even for the points that are not non or inf.

Example code:

`# This entire item doesn't get drawn because it contains nan:`
`line = PlotCurveItem(connect='finite')`
`line.setData(x=[1,2,numpy.nan,3,4], y=[1,2,numpy.nan,3,4])`

`# This entire item doesn't get drawn because it contains inf:`
`line = PlotCurveItem(connect='finite')`
`line.setData(x=[1,2,numpy.inf,3,4], y=[1,2,numpy.inf,3,4])`

`# This item does get drawn on the chart because it has neither nan nor inf:`
`line = PlotCurveItem(connect='finite')`
`line.setData(x=[1,2,3,4], y=[1,2,3,4])`

If I can't insert nan or inf to break the line into segments, then what is the point of "connect='finite'"? And if there's a new way of doing this, how can I break the line into segments alternatively?

Thank you,

jy92lee

edit: When I first submitted this post, I thought it was due to recent changes made in pyqtgraph's latest version. But upon further experiments, I came to a conclusion that this bug only occurs when I use pyqtgraph installed from conda. It works fine when I install pyqtgraph from pypi.

**Comment by pijyoi** [comment]
It seems to work for me. The keyword arguments are `x`, `y` and not `xs`, `ys`. No error is given for the unrecognized keyword arguments.
```python
import pyqtgraph as pg
import numpy

pg.mkQApp()

pw = pg.PlotWidget()
pw.show()

line = pg.PlotCurveItem(connect='finite')
line.setData(x=[1,2,numpy.nan,3,4], y=[1,2,numpy.nan,3,4])
pw.addItem(line)

line = pg.PlotCurveItem(connect='finite')
line.setData(x=[1,2,numpy.inf,3,4], y=[2,3,numpy.inf,4,5])
pw.addItem(line)

line = pg.PlotCurveItem(connect='finite')
line.setData(x=[1,2,3,4], y=[3,4,5,6])
pw.addItem(line)

pg.exec()
```

**Reply by proptrader33** [reply]
You're right, they should be x and y, not xs and ys. I didn't verfy the correctness of the code. But it still didn't work for me even after using the right parameter names.

Upon further analysis I came to realize that this bug occurs only when I usepyqtgraph installed from conda (https://anaconda.org/anaconda/pyqtgraph). I updated conda to the latest version and PlotCurveItems still didn't work. Then I switched to the pypi version (https://pypi.org/project/pyqtgraph/) and it works perfectly now.

Strange how it was working perfectly fine a few hours before writing this thread yesterday and then suddenly stopped working.

**Comment by j9ac9k** [comment]
Which Qt bindings are you using?

**Reply by proptrader33** [reply]
pyqt 5

But as I said in the above response to another person, I have now switched to pypi's pyqtgraph and no longer have this problem.

This bug does not occur when I use pyqtgraph installed from pypi. It's only a problem when I use pyqtgraph installed from conda. I had been using pyqtgraph for 2-3 years from conda with no such problem until yesterday. I don't know why it suddenly stopped working yesterday.

**Reply by j9ac9k** [reply]
pyqt5 via conda is not a supported configuration (it uses PyQt 5.9, we mandate Qt 5.12+).  If installing Qt bindings via conda, you need to use the pyside2 bindings, if both are installed, pyqtgraph may attempt to use the PyQt5 ones; but you should get a bit of a warning about that.

If you've installed PyQt5 from pypi that will work fine.  Can you give the output from `conda list`  and `pip list` ?

**Reply by proptrader33** [reply]
Here's my `conda list`:

`# Name                    Version                   Build  Channel`
`_libgcc_mutex             0.1                        main`
`_openmp_mutex             4.5                       1_gnu`
`aiohttp                   3.8.1                    pypi_0    pypi`
`aiosignal                 1.2.0                    pypi_0    pypi`
`async-timeout             4.0.2                    pypi_0    pypi`
`attrs                     21.4.0                   pypi_0    pypi`
`backports                 1.1                pyhd3eb1b0_0`
`backports.zoneinfo        0.2.1            py38h497a2fe_3    conda-forge`
`brotli                    1.0.9                he6710b0_2`
`brotlipy                  0.7.0           py38h27cfd23_1003`
`ca-certificates           2022.4.26            h06a4308_0`
`certifi                   2021.10.8        py38h06a4308_2`
`cffi                      1.15.0           py38hd667e15_1`
`chardet                   3.0.4           py38h06a4308_1003`
`charset-normalizer        2.0.12                   pypi_0    pypi`
`cryptography              36.0.0           py38h9ce1e76_0`
`cycler                    0.11.0             pyhd3eb1b0_0`
`cython                    0.29.24          py38h709712a_0    conda-forge`
`dateparser                1.1.0                    pypi_0    pypi`
`dbus                      1.13.18              hb2f20db_0`
`expat                     2.4.4                h295c915_0`
`fontconfig                2.13.1               h6c09931_0`
`fonttools                 4.25.0             pyhd3eb1b0_0`
`freetype                  2.11.0               h70c0345_0`
`frozenlist                1.3.0                    pypi_0    pypi`
`giflib                    5.2.1                h7b6447c_0`
`glib                      2.69.1               h4ff587b_1`
`gst-plugins-base          1.14.0               h8213a91_2`
`gstreamer                 1.14.0               h28cd5cc_2`
`icu                       58.2                 he6710b0_3`
`idna                      2.10               pyhd3eb1b0_0`
`jpeg                      9e                   h7f8727e_0`
`kiwisolver                1.3.2            py38h295c915_0`
`lcms2                     2.12                 h3be6417_0`
`ld_impl_linux-64          2.35.1               h7274673_9`
`libblas                   3.9.0           11_linux64_openblas    conda-forge`
`libcblas                  3.9.0           11_linux64_openblas    conda-forge`
`libffi                    3.3                  he6710b0_2`
`libgcc-ng                 9.3.0               h5101ec6_17`
`libgfortran-ng            7.5.0               ha8ba4b0_17`
`libgfortran4              7.5.0               ha8ba4b0_17`
`libgomp                   9.3.0               h5101ec6_17`
`liblapack                 3.9.0           11_linux64_openblas    conda-forge`
`libopenblas               0.3.17               hf726d26_1`
`libpng                    1.6.37               hbc83047_0`
`libsodium                 1.0.18               h7b6447c_0`
`libstdcxx-ng              9.3.0               hd4cf53a_17`
`libtiff                   4.2.0                h85742a9_0`
`libuuid                   1.0.3                h7f8727e_2`
`libwebp                   1.2.2                h55f646e_0`
`libwebp-base              1.2.2                h7f8727e_0`
`libxcb                    1.14                 h7b6447c_0`
`libxml2                   2.9.12               h03d6c58_0`
`lz4-c                     1.9.3                h295c915_1`
`matplotlib                3.5.1            py38h06a4308_1`
`matplotlib-base           3.5.1            py38ha18d171_1`
`multidict                 6.0.2                    pypi_0    pypi`
`munkres                   1.1.4                      py_0`
`ncurses                   6.3                  h7f8727e_2`
`numpy                     1.20.3           py38h9894fe3_1    conda-forge`
`openssl                   1.1.1n               h7f8727e_0`
`packaging                 21.3               pyhd3eb1b0_0`
`pandas                    1.4.0                    pypi_0    pypi`
`pcre                      8.45                 h295c915_0`
`pillow                    9.0.1            py38h22f2fdc_0`
`pip                       21.2.4           py38h06a4308_0`
`pycparser                 2.21               pyhd3eb1b0_0`
`pyopenssl                 22.0.0             pyhd3eb1b0_0`
`pyparsing                 3.0.4              pyhd3eb1b0_0`
`pyqt                      5.9.2            py38h05f1152_4`
`pyqt5                     5.15.4                   pypi_0    pypi`
`pyqt5-qt5                 5.15.2                   pypi_0    pypi`
`pyqt5-sip                 12.9.0                   pypi_0    pypi`
`pyqtgraph                 0.12.4                   pypi_0    pypi`
`pysocks                   1.7.1            py38h06a4308_0`
`python                    3.8.5                h7579374_1    anaconda`
`python-binance            1.0.15                   pypi_0    pypi`
`python-dateutil           2.8.2              pyhd3eb1b0_0`
`python_abi                3.8                      2_cp38    conda-forge`
`pytz                      2020.1                     py_0    anaconda`
`pyzmq                     19.0.2           py38ha71036d_2    conda-forge`
`qt                        5.9.7                h5867ecd_1`
`readline                  8.1.2                h7f8727e_1`
`regex                     2022.3.2                 pypi_0    pypi`
`requests                  2.24.0                     py_0    anaconda`
`setuptools                61.2.0           py38h06a4308_0`
`simpleaudio               1.0.4                    pypi_0    pypi`
`sip                       4.19.13          py38h295c915_0`
`six                       1.16.0             pyhd3eb1b0_1`
`speechrecognition         3.8.1                    pypi_0    pypi`
`sqlite                    3.38.3               hc218d9a_0`
`tk                        8.6.11               h1ccaba5_0`
`tornado                   6.1              py38h27cfd23_0`
`tzdata                    2022a                hda174b7_0`
`tzlocal                   3.0              py38h578d9bd_2    conda-forge`
`ujson                     5.1.0                    pypi_0    pypi`
`urllib3                   1.25.11                    py_0`
`websockets                9.1                      pypi_0    pypi`
`wheel                     0.37.1             pyhd3eb1b0_0`
`xz                        5.2.5                h7f8727e_1`
`yarl                      1.7.2                    pypi_0    pypi`
`zeromq                    4.3.4                h2531618_0`
`zlib                      1.2.12               h7f8727e_2`
`zstd                      1.4.9                haebb681_0`

And here's my `pip list`:

`Package            Version`
`------------------ ---------`
`aiohttp            3.8.1`
`aiosignal          1.2.0`
`async-timeout      4.0.2`
`attrs              21.4.0`
`backports.zoneinfo 0.2.1`
`brotlipy           0.7.0`
`certifi            2021.10.8`
`cffi               1.15.0`
`chardet            3.0.4`
`charset-normalizer 2.0.12`
`cryptography       36.0.0`
`cycler             0.11.0`
`Cython             0.29.24`
`dateparser         1.1.0`
`fonttools          4.25.0`
`frozenlist         1.3.0`
`idna               2.10`
`kiwisolver         1.3.2`
`matplotlib         3.5.1`
`multidict          6.0.2`
`munkres            1.1.4`
`numpy              1.20.3`
`packaging          21.3`
`pandas             1.4.0`
`Pillow             9.0.1`
`pip                21.2.4`
`pycparser          2.21`
`pyOpenSSL          22.0.0`
`pyparsing          3.0.4`
`PyQt5              5.15.4`
`PyQt5-Qt5          5.15.2`
`PyQt5-sip          12.9.0`
`pyqtgraph          0.12.4`
`PySocks            1.7.1`
`python-binance     1.0.15`
`python-dateutil    2.8.2`
`pytz               2020.1`
`pyzmq              19.0.2`
`regex              2022.3.2`
`requests           2.24.0`
`setuptools         61.2.0`
`simpleaudio        1.0.4`
`sip                4.19.13`
`six                1.16.0`
`SpeechRecognition  3.8.1`
`tornado            6.1`
`tzlocal            3.0`
`ujson              5.1.0`
`urllib3            1.25.11`
`websockets         9.1`
`wheel              0.37.1`
`yarl               1.7.2`

**Reply by j9ac9k** [reply]
Looks like you have multiple versions of PyQt5 installed.  For the record, we don't support PyQt5 5.9, some things may work, but it was a lot more trouble than it was worth so we dropped support with the release of 0.12.0.

```
pyqt 5.9.2 py38h05f1152_4
pyqt5 5.15.4 pypi_0 pypi
pyqt5-qt5 5.15.2 pypi_0 pypi
...
qt 5.9.7 h5867ecd_1
```

I would suggest `conda uninstall pyqt5`  or if you want to stick w/ a conda distribution of pyside bindings `conda install -c conda-forge pyside2` however those are quite old.
